File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        if(grid == null || grid.length == 0){
            return 0;
        }

        int freshOranges = 0;
        int mins = 0;
        Queue<int []> q = new LinkedList<>();

        int m = grid.length; //row
        int n = grid[0].length; // col

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j ++){
                if(grid[i][j] == 2){
                    q.add(new int[]{i, j});
                }
                else if (grid[i][j] == 1){
                    freshOranges += 1;
                }
            }
        }

        if(freshOranges == 0){
            return 0;
        }

        int[][] dirs = {{0,1}, {1,0}, {-1,0}, {0,-1}};

        while(!q.isEmpty()){
            int size = q.size();

            for(int i =0; i < size; i++){
                int[] currOrange = q.poll();

                for(int[] dir : dirs){
                    int nr = dir[0] + currOrange[0];
                    int nc = dir[1] + currOrange[1];

                    if((nr >= 0 && nr < m) && (nc >= 0 && nc < n) && (grid[nr][nc] == 1)){
                        grid[nr][nc] = 2;
                        q.add(new int[] {nr, nc});
                        freshOranges -= 1;
                    }
                }
            }
            mins += 1;
        }

        if(freshOranges == 0){
            return mins - 1;
        }
        return -1;

    }
}

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        time = [[math.inf] * cols for i in range(rows)]

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 2:
                    self.dfs(grid, time, i, j, 0)
        
        max_time = 0

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1 and time[i][j] == math.inf:
                    # print(time)
                    return -1
                if time[i][j] != math.inf:
                    max_time = max(max_time, time[i][j])
        # print(time)
        return max_time
    
    def dfs(self, grid, time , row, col, current_time):
        rows = len(grid)
        cols = len(grid[0])

        #boudary check
        if row < 0 or row >= rows or col < 0 or col >= cols:
            return 
        
        #if empty cell
        if grid[row][col] == 0:
            return
        
        #if already visited with smaller time -> stop
        if time[row][col] <= current_time:
            return
        
        time[row][col] = current_time

        #explore directions
        dirs = [(0,1), (0,-1), (1,0), (-1,0)]

        for dr, dc in dirs:
            self.dfs(grid, time, row + dr, col + dc, current_time + 1)
